metrics/user_registries.go

```diff
@@ -0,0 +1,811 @@
+	}
+}
+
+// MetricFamilyMap is a map of metric names to their family (metrics with same name, but different labels)
+// Keeping map of metric name to its family makes it easier to do searches later.
+type MetricFamilyMap map[string]*dto.MetricFamily
+
+// NewMetricFamilyMapFromGatherer is like NewMetricFamilyMap but gets metrics directly from a
+// prometheus.Gatherer instance.
+func NewMetricFamilyMapFromGatherer(gatherer prometheus.Gatherer) (MetricFamilyMap, error) {
+	metrics, err := gatherer.Gather()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewMetricFamilyMap(metrics)
+}
+
+// NewMetricFamilyMap sorts output from Gatherer.Gather method into a map.
+// Gatherer.Gather specifies that there metric families are uniquely named, and we use that fact here.
+// If they are not, this method returns error.
+func NewMetricFamilyMap(metrics []*dto.MetricFamily) (MetricFamilyMap, error) {
+	perMetricName := MetricFamilyMap{}
+
+	for _, m := range metrics {
+		name := m.GetName()
+		// these errors should never happen when passing Gatherer.Gather() output.
+		if name == "" {
+			return nil, errors.New("empty name for metric family")
+		}
+		if perMetricName[name] != nil {
+			return nil, fmt.Errorf("non-unique name for metric family: %q", name)
+		}
+
+		perMetricName[name] = m
+	}
+
+	return perMetricName, nil
+}
+
+func (mfm MetricFamilyMap) SumCounters(name string) float64 {
+	return sum(mfm[name], counterValue)
+}
+
+func (mfm MetricFamilyMap) SumGauges(name string) float64 {
+	return sum(mfm[name], gaugeValue)
+}
+
+func (mfm MetricFamilyMap) MaxGauges(name string) float64 {
+	return max(mfm[name], gaugeValue)
+}
+
+func (mfm MetricFamilyMap) SumHistograms(name string) HistogramData {
+	hd := HistogramData{}
+	mfm.SumHistogramsTo(name, &hd)
+	return hd
+}
+
+func (mfm MetricFamilyMap) SumHistogramsTo(name string, output *HistogramData) {
+	for _, m := range mfm[name].GetMetric() {
+		output.AddHistogram(m.GetHistogram())
+	}
+}
+
+func (mfm MetricFamilyMap) SumSummaries(name string) SummaryData {
+	sd := SummaryData{}
+	mfm.SumSummariesTo(name, &sd)
+	return sd
+}
+
+func (mfm MetricFamilyMap) SumSummariesTo(name string, output *SummaryData) {
+	for _, m := range mfm[name].GetMetric() {
+		output.AddSummary(m.GetSummary())
+	}
+}
+
+func (mfm MetricFamilyMap) sumOfSingleValuesWithLabels(metric string, labelNames []string, extractFn func(*dto.Metric) float64, aggregateFn func(labelsKey string, labelValues []string, value float64), skipZeroValue bool) {
+	metricsPerLabelValue := getMetricsWithLabelNames(mfm[metric], labelNames)
+
+	for key, mlv := range metricsPerLabelValue {
+		for _, m := range mlv.metrics {
+			val := extractFn(m)
+			if skipZeroValue && val == 0 {
+				continue
+			}
+
+			aggregateFn(key, mlv.labelValues, val)
+		}
+	}
+}
+
+// MetricFamiliesPerUser is a collection of metrics gathered via calling Gatherer.Gather() method on different
+// gatherers, one per user.
+type MetricFamiliesPerUser []struct {
```

Comment:
Could we use `Tenant` in place of `User` here since we have a `tenant/` package in dskit and use that term elsewhere?